res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+    PioneerDDJ400.loopAdjustIn[channel] = false;
+};
+
+PioneerDDJ400.setReloopLight = function(status, value) {
+    midi.sendShortMsg(status, 0x4D, value);
+    midi.sendShortMsg(status, 0x50, value);
+};
+
+
+PioneerDDJ400.setLoopButtonLights = function(status, value) {
+    [0x10, 0x11, 0x4E, 0x4C].forEach(function(control) {
+        midi.sendShortMsg(status, control, value);
+    });
+};
+
+PioneerDDJ400.startLoopLightsBlink = function(channel, control, status, group) {
+    var blink = 0x7F;
+
+    PioneerDDJ400.stopLoopLightsBlink(group, control, status);
+
+    PioneerDDJ400.timers[group][control] = engine.beginTimer(500, function() {
+        blink = 0x7F - blink;
+
+        // When adjusting the loop out position, turn the loop in light off
+        if (PioneerDDJ400.loopAdjustOut[channel]) {
+            midi.sendShortMsg(status, 0x10, 0x00);
+            midi.sendShortMsg(status, 0x4C, 0x00);
+        } else {
+            midi.sendShortMsg(status, 0x10, blink);
+            midi.sendShortMsg(status, 0x4C, blink);
+        }
+
+        // When adjusting the loop in position, turn the loop out light off
+        if (PioneerDDJ400.loopAdjustIn[channel]) {
+            midi.sendShortMsg(status, 0x11, 0x00);
+            midi.sendShortMsg(status, 0x4E, 0x00);
+        } else {
+            midi.sendShortMsg(status, 0x11, blink);
+            midi.sendShortMsg(status, 0x4E, blink);
+        }
+    });
+
+};
+
+PioneerDDJ400.stopLoopLightsBlink = function(group, control, status) {
+    PioneerDDJ400.timers[group] = PioneerDDJ400.timers[group] || {};
+
+    if (PioneerDDJ400.timers[group][control] !== undefined) {
+        engine.stopTimer(PioneerDDJ400.timers[group][control]);
+    }
+    PioneerDDJ400.timers[group][control] = undefined;
+    PioneerDDJ400.setLoopButtonLights(status, 0x7F);
+};
+
+PioneerDDJ400.loopToggle = function(value, group, control) {
+    var status = group === "[Channel1]" ? 0x90 : 0x91,
+        channel = group === "[Channel1]" ? 0 : 1;
+
+    PioneerDDJ400.setReloopLight(status, value ? 0x7F : 0x00);
+
+    if (value) {
+        PioneerDDJ400.startLoopLightsBlink(channel, control, status, group);
+    } else {
+        PioneerDDJ400.stopLoopLightsBlink(group, control, status);
+        PioneerDDJ400.loopAdjustOut[channel] = PioneerDDJ400.loopAdjustIn[channel] = false;
+    }
+};
+
+//
+// CUE/LOOP CALL
+//
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group) {
+    if (value) {
+        engine.setValue(group, "loop_scale", 0.5);
+    }
+};
+
+PioneerDDJ400.cueLoopCallRight = function(_channel, _control, value, _status, group) {
+    if (value) {
+        engine.setValue(group, "loop_scale", 2.0);
+    }
+};
+
+//
+// BEAT SYNC
+//
+PioneerDDJ400.syncPressed = function(channel, control, value, status, group) {
+    engine.setValue(group, "sync_enabled", value);
+};
+
+PioneerDDJ400.syncLongPressed = function(channel, control, value, status, group) {
```

Comment:
I'm quite confused by this. The controller sends a different MIDI message when the button is long pressed? I've never heard of a controller doing that.

Reply:
Yes, it really does. Makes implementing sync counterintuitive.

Reply:
Okay, please add a comment explaining that.